PuLP_integration_example/Sudoku_pulp.ipynb

```diff
@@ -0,0 +1,220 @@
+      ],
+      "metadata": {
+        "id": "mIX0AKStL4It"
+      }
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "lQdRtS9Bjdym"
+      },
+      "outputs": [],
+      "source": [
+        "\"\"\"\n",
+        "The Sudoku Problem Formulation for the PuLP Modeller\n",
+        "\n",
+        "Authors: Antony Phillips, Dr Stuart Mitchell\n",
+        "edited by Nathan Sudermann-Merx\n",
+        "\"\"\"\n",
+        "\n",
+        "# Import PuLP modeler functions\n",
+        "from pulp import *\n",
+        "\n",
+        "# All rows, columns and values within a Sudoku take values from 1 to 9\n",
+        "VALS = ROWS = COLS = range(1, 10)\n",
+        "\n",
+        "# The boxes list is created, with the row and column index of each square in each box\n",
+        "Boxes = [\n",
+        "    [(3 * i + k + 1, 3 * j + l + 1) for k in range(3) for l in range(3)]\n",
+        "    for i in range(3)\n",
+        "    for j in range(3)\n",
+        "]\n",
+        "\n",
+        "# The prob variable is created to contain the problem data\n",
+        "prob = LpProblem(\"Sudoku Problem\")\n",
+        "\n",
+        "# The decision variables are created\n",
+        "choices = LpVariable.dicts(\"Choice\", (VALS, ROWS, COLS), cat=\"Binary\")\n",
+        "\n",
+        "# We do not define an objective function since none is needed\n",
+        "\n",
+        "# A constraint ensuring that only one value can be in each square is created\n",
+        "for r in ROWS:\n",
+        "    for c in COLS:\n",
+        "        prob += lpSum([choices[v][r][c] for v in VALS]) == 1\n",
+        "\n",
+        "# The row, column and box constraints are added for each value\n",
+        "for v in VALS:\n",
+        "    for r in ROWS:\n",
+        "        prob += lpSum([choices[v][r][c] for c in COLS]) == 1\n",
+        "\n",
+        "    for c in COLS:\n",
+        "        prob += lpSum([choices[v][r][c] for r in ROWS]) == 1\n",
+        "\n",
+        "    for b in Boxes:\n",
+        "        prob += lpSum([choices[v][r][c] for (r, c) in b]) == 1\n",
+        "\n",
+        "# The starting numbers are entered as constraints\n",
+        "input_data = [\n",
```

Comment:
Lets add details how these numbers are added 
```
The starting numbers are entered as constraints i.e a 5 in row 1 column 1 is true.
```
And also mention the indices start from 1 - 9